### PR TITLE
Make combine_similar less greedy for merge

### PR DESCRIPTION
**Pull request by phofl**
This makes combine_similar less greedy, we are currently removing projections on both sides and see if we are equal, we should do one after another to avoid pushing projections up unnecessarily. Shuffling is expensive, so this should help quite a bit.

This sits on top of #333, the previous structure made this change even harder that it was anyway

### CODE REVIEW

File: dask_expr/_expr.py [hunk 1]
```diff
@@ -417,7 +417,12 @@ def combine_similar(
 
             if changed_dependency:
                 expr = type(expr)(*new_operands)
-                changed = True
+                if isinstance(expr, Projection):
+                    # We might introduce stacked Projections (merge for example).
```
Comment (phofl) [inline dask_expr/_expr.py:421]: Merge potentially introduces multiple Projections based on how the tree is constructed, we should squash those together again.

I am special casing for Projections for now, but I think there is no harm in opening this up to all Expressions as long as we don't call simplify_up. But I want more test coverage for this, so not doing it here

File: dask_expr/tests/test_merge.py [hunk 2]
```diff
@@ -206,3 +207,26 @@ def test_merge_combine_similar(npartitions_left, npartitions_right):
     expected["new"] = expected.b + expected.c
     expected = expected.groupby(["a", "e", "x"]).new.sum()
     assert_eq(query, expected)
+
+
+def test_merge_combine_similar_intermediate_projections():
```
Comment (rjzamora) [inline dask_expr/tests/test_merge.py:212]: Is there a reproducer for the behavior being targeted by this PR? This test seems to pass on `main` for me.

Comment (phofl) [reply dask_expr/tests/test_merge.py:212]: Yeah the bug was blocked by the additional layer in #333 but shows up now

**Comment by phofl** [comment]
Previous tree with the reproducer:

```
Assign: key='new'
  Projection: columns=['b', 'x', 'y']
    Projection: columns=['b', 'x', 'd', 'y']
      HashJoinP2P: left_on='b' right_on='d' left_index=False right_index=False shuffle_left_on='b' shuffle_right_on='d'
        HashJoinP2P: left_on=['a'] right_on=['a'] left_index=False right_index=False shuffle_left_on=['a'] shuffle_right_on=['a']
          FromPandas: frame='<pandas>' npartitions=2 columns=['a', 'b']
          FromPandas: frame='<pandas>' npartitions=3
        FromPandas: frame='<pandas>' npartitions=3 columns=['d', 'y']
  Add:
    Projection: columns='b'
      Projection: columns=['b', 'a', 'd']
        HashJoinP2P: left_on='b' right_on='d' left_index=False right_index=False shuffle_left_on='b' shuffle_right_on='d'
          HashJoinP2P: left_on=['a'] right_on=['a'] left_index=False right_index=False shuffle_left_on=['a'] shuffle_right_on=['a']
            FromPandas: frame='<pandas>' npartitions=2 columns=['a', 'b']
            FromPandas: frame='<pandas>' npartitions=3
          FromPandas: frame='<pandas>' npartitions=3 columns=['d', 'y']
    Projection: columns='x'
      Projection: columns=['b', 'x', 'd']
        HashJoinP2P: left_on='b' right_on='d' left_index=False right_index=False shuffle_left_on='b' shuffle_right_on='d'
          HashJoinP2P: left_on=['a'] right_on=['a'] left_index=False right_index=False shuffle_left_on=['a'] shuffle_right_on=['a']
            FromPandas: frame='<pandas>' npartitions=2 columns=['a', 'b']
            FromPandas: frame='<pandas>' npartitions=3
          FromPandas: frame='<pandas>' npartitions=3 columns=['d', 'y']
```

We have the 2 Join layers right next to each other without dropping unnecessary columns

The new tree:

```
Assign: key='new'
  Projection: columns=['b', 'x', 'y']
    HashJoinP2P: left_on='b' right_on='d' left_index=False right_index=False shuffle_left_on='b' shuffle_right_on='d'
      Projection: columns=['b', 'x']
        HashJoinP2P: left_on=['a'] right_on=['a'] left_index=False right_index=False shuffle_left_on=['a'] shuffle_right_on=['a']
          FromPandas: frame='<pandas>' npartitions=2 columns=['a', 'b']
          FromPandas: frame='<pandas>' npartitions=3
      FromPandas: frame='<pandas>' npartitions=3 columns=['d', 'y']
  Add:
    Projection: columns='b'
      HashJoinP2P: left_on='b' right_on='d' left_index=False right_index=False shuffle_left_on='b' shuffle_right_on='d'
        Projection: columns=['b']
          HashJoinP2P: left_on=['a'] right_on=['a'] left_index=False right_index=False shuffle_left_on=['a'] shuffle_right_on=['a']
            FromPandas: frame='<pandas>' npartitions=2 columns=['a', 'b']
            Projection: columns=['a']
              FromPandas: frame='<pandas>' npartitions=3
        Projection: columns=['d']
          FromPandas: frame='<pandas>' npartitions=3 columns=['d', 'y']
    Projection: columns='x'
      HashJoinP2P: left_on='b' right_on='d' left_index=False right_index=False shuffle_left_on='b' shuffle_right_on='d'
        Projection: columns=['b', 'x']
          HashJoinP2P: left_on=['a'] right_on=['a'] left_index=False right_index=False shuffle_left_on=['a'] shuffle_right_on=['a']
            FromPandas: frame='<pandas>' npartitions=2 columns=['a', 'b']
            FromPandas: frame='<pandas>' npartitions=3
        Projection: columns=['d']
          FromPandas: frame='<pandas>' npartitions=3 columns=['d', 'y']
```

We have an intermediate Projection that drops unnecessary columns right away

**Review by rjzamora — APPROVED** [review]
This logic is pretty hard to follow (mostly because it was **already** complicated), but I don't have any better ideas, and I don't think it is worth the effort to spend much time attempting a simplification until we have addressed https://github.com/dask-contrib/dask-expr/issues/336

File: dask_expr/_expr.py (hunk 1 above)
Comment (phofl) [reply dask_expr/_expr.py:421]: It makes follow up steps more complicated if you have to start worrying about multiple projections stacked upon each other. This is more to reduce mental load than anything else. (It would have made parts of the combine_similar logic even more complicated).

Performance impact should ne negligible indeed.

Comment (rjzamora) [reply dask_expr/_expr.py:421]: Okay, makes sense to me. Thanks for clarifying.

File: dask_expr/_merge.py [hunk 3]
```diff
@@ -23,6 +23,7 @@
 from dask_expr._util import _convert_to_list
 
 _HASH_COLUMN_NAME = "__hash_partition"
+_PARTITION_COLUMNS = "_partitions"
```
[the remark this answers is not in the document]
Comment (phofl) [reply dask_expr/_merge.py]: Oh good point, thx
